airbyte-integrations/connectors/source-shopify/source_shopify/utils.py

```diff
@@ -323,3 +325,38 @@ def decorator(*args, **kwargs) -> Any:
             return func(*args, **kwargs)
 
         return decorator
+
+
+class LimitReducingErrorHandler(HttpStatusErrorHandler):
+    """
+    Error handler that halves the page size (limit) on each 500 error, down to 1.
+    No stream instance required; operates directly on the request URL.
+    """
+
+    def __init__(self, max_retries: int, error_mapping: dict):
+        super().__init__(logger=None, max_retries=max_retries, error_mapping=error_mapping)
+
+    def interpret_response(self, response_or_exception):
```

Comment:
can we add a type hint for `response_or_exception`? Or actually just make it match the signature of `HttpStatusErrorHandler`. The hint should be Optional[Union[requests.Response, Exception]] = None. And we should have a return type